// _anyproxy@4.0.0-beta15@anyproxy/lib/rule_default.js
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = {

  summary: 'the default rule for AnyProxy',

  /**
   * 
   * 
   * @param {object} requestDetail
   * @param {string} requestDetail.protocol
   * @param {object} requestDetail.requestOptions
   * @param {object} requestDetail.requestData
   * @param {object} requestDetail.response
   * @param {number} requestDetail.response.statusCode
   * @param {object} requestDetail.response.header
   * @param {buffer} requestDetail.response.body
   * @returns
   */
  beforeSendRequest: /*#__PURE__*/_regenerator2.default.mark(function beforeSendRequest(requestDetail) {
    return _regenerator2.default.wrap(function beforeSendRequest$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt('return', null);

          case 1:
          case 'end':
            return _context.stop();
        }
      }
    }, beforeSendRequest, this);
  }),


  /**
   * 
   * 
   * @param {object} requestDetail
   * @param {object} responseDetail
   */
  beforeSendResponse: /*#__PURE__*/_regenerator2.default.mark(function beforeSendResponse(requestDetail, responseDetail) {
    return _regenerator2.default.wrap(function beforeSendResponse$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt('return', null);

          case 1:
          case 'end':
            return _context2.stop();
        }
      }
    }, beforeSendResponse, this);
  }),


  /**
   * 
   * 
   * @param {any} requestDetail 
   * @returns 
   */
  beforeDealHttpsRequest: /*#__PURE__*/_regenerator2.default.mark(function beforeDealHttpsRequest(requestDetail) {
    return _regenerator2.default.wrap(function beforeDealHttpsRequest$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            return _context3.abrupt('return', false);

          case 1:
          case 'end':
            return _context3.stop();
        }
      }
    }, beforeDealHttpsRequest, this);
  }),


  /**
   * 
   * 
   * @param {any} requestDetail 
   * @param {any} error 
   * @returns 
   */
  onError: /*#__PURE__*/_regenerator2.default.mark(function onError(requestDetail, error) {
    return _regenerator2.default.wrap(function onError$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            return _context4.abrupt('return', null);

          case 1:
          case 'end':
            return _context4.stop();
        }
      }
    }, onError, this);
  }),


  /**
   * 
   * 
   * @param {any} requestDetail 
   * @param {any} error 
   * @returns 
   */
  onConnectError: /*#__PURE__*/_regenerator2.default.mark(function onConnectError(requestDetail, error) {
    return _regenerator2.default.wrap(function onConnectError$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            return _context5.abrupt('return', null);

          case 1:
          case 'end':
            return _context5.stop();
        }
      }
    }, onConnectError, this);
  })
};
